### PR TITLE
datetime: fix cases when datetime nanoseconds could be represented incorrectly

### LINKED ISSUE #8570: Datetime after nsec subtration weird string representation
**Bug description**

If nanoseconds subtracted from a datetime with fewer nanoseconds than in interval, new datetime string representation contains negative nanoseconds. They are exported to msgpack as negatives as well.

Tarantool 2.10.6-0-g3990f976b

**Steps to reproduce + Actual behavior**
```
tarantool> dt = datetime.new{year=2008, month=2, day=3, hour=3, min=36, sec=43}
---
...

tarantool> intv = datetime.interval.new{nsec=10000023}
---
...

tarantool> dt - intv
---
- 2008-02-03T03:36:43.-10000023Z
...
```

**Expected behavior**

```
tarantool> dt - intv
---
- 2008-02-03T03:36:42.989999977Z
...
```

### PR DESCRIPTION
**Pull request by olegrok**
Closes #8570

### CODE REVIEW

File: changelogs/unreleased/gh-8570-fix-datetime-str-negative-nsec.md
Comment (andreyaksenov) [inline changelogs/unreleased/gh-8570-fix-datetime-str-negative-nsec.md:4]: Added few articles:
```
* Fixed errors when the string representation of a datetime object had
a negative nanosecond part (gh-8570).
```

**Review by andreyaksenov — APPROVED** [review]

File: changelogs/unreleased/gh-8570-fix-datetime-str-negative-nsec.md
Comment (olegrok) [reply changelogs/unreleased/gh-8570-fix-datetime-str-negative-nsec.md]: Applied. Thanks.

File: src/lua/datetime.lua
Comment (ligurio) [inline src/lua/datetime.lua:1092]: add a comment here with explanation

**Review by ligurio — COMMENTED** [review]
Thanks for the patch, see my comments.

**Review by ligurio — COMMENTED** [review]
Thanks for the patch, see my comments.

File: test/app-tap/datetime.test.lua
Comment (olegrok) [reply test/app-tap/datetime.test.lua]:
```diff
diff --git a/test/app-tap/datetime.test.lua b/test/app-tap/datetime.test.lua
index 5bdd94793f..790ec29ac6 100755
--- a/test/app-tap/datetime.test.lua
+++ b/test/app-tap/datetime.test.lua
@@ -222,8 +222,8 @@ test:test("Simple date creation by attributes", function(test)
             '2021-08-30T21:31:11.000123Z', '{timestamp.usec}')
     test:is(tostring(date.new{timestamp = 1630359071, nsec = 123}),
             '2021-08-30T21:31:11.000000123Z', '{timestamp.nsec}')
-    test:is(tostring(date.new{timestamp = -192799798.5}),
-            '1963-11-22T12:30:01.500Z', '{negative timestamp}')
+    test:is(tostring(date.new{timestamp = -0.1}),
+            '1969-12-31T23:59:59.900Z', '{negative timestamp}')
 end)

test:test("Simple date creation by attributes - check failed", function(test)
@@ -1815,8 +1815,8 @@ test:test("Time :set{} operations", function(test)
             '2021-08-30T21:31:11.000123+0800', 'timestamp + usec')
     test:is(tostring(ts:set{timestamp = 1630359071, nsec = 123}),
             '2021-08-30T21:31:11.000000123+0800', 'timestamp + nsec')
-    test:is(tostring(ts:set{timestamp = -192799798.5}),
-            '1963-11-22T12:30:01.500+0800', 'negative timestamp')
+    test:is(tostring(ts:set{timestamp = -0.1}),
+            '1969-12-31T23:59:59.900+0800', 'negative timestamp')
 end)
 
 test:test("Check :set{} and .new{} equal for all attributes", function(test)
@@ -1850,7 +1850,7 @@ test:test("Check :set{} and .new{} equal for all attributes", function(test)
     test:is(ts, ts2, ('timestamp+tzoffset (%s = %s)'):
             format(tostring(ts), tostring(ts2)))
 
-    obj = {timestamp = -192799798.5, tzoffset = '+0800'}
+    obj = {timestamp = -0.1, tzoffset = '+0800'}
     ts = date.new(obj)
     ts2 = date.new():set(obj)
     test:is(ts, ts2, ('negative timestamp+tzoffset (%s = %s)'):
```

File: src/lua/datetime.lua
Comment (olegrok) [reply src/lua/datetime.lua:1098]: Added. (With simple changes to be consistent in assign operations)

```diff
--- a/src/lua/datetime.lua
+++ b/src/lua/datetime.lua
@@ -588,9 +588,12 @@ local function datetime_new(obj)
         end
         local fraction
         s, fraction = math_modf(ts)
+        -- In case of negative fraction part we should
+        -- make it positive at the expense of the integer part.
+        -- Code below expects that "nsec" value is always positive.
         if fraction < 0 then
             s = s - 1
-            fraction = 1 + fraction
+            fraction = fraction + 1
         end
         -- if there are separate nsec, usec, or msec provided then
         -- timestamp should be integer
@@ -1089,9 +1092,12 @@ local function datetime_set(self, obj)
         end
         local sec_int, fraction
         sec_int, fraction = math_modf(ts)
+        -- In case of negative fraction part we should
+        -- make it positive at the expense of the integer part.
+        -- Code below expects that "nsec" value is always positive.
         if fraction < 0 then
             sec_int = sec_int - 1
-            fraction = 1 + fraction
+            fraction = fraction + 1
         end
```

**Comment by olegrok** [comment]
CI failure seems to be not related to my changes:
```
2023-05-12T11:58:39.0238771Z [018] Test failed! Output from reject file /tmp/t/rejects/vinyl-luatest/select_consistency.reject:
2023-05-12T11:58:39.0240260Z [018] Tarantool version is 3.0.0-entrypoint-220-gbb79e5cad
2023-05-12T11:58:39.0241299Z [018] TAP version 13
2023-05-12T11:58:39.0394915Z [018] 1..2
2023-05-12T11:58:39.0395948Z [018] # Started on Fri May 12 11:56:33 2023
2023-05-12T11:58:39.0396849Z [018] # Starting group: vinyl.select_consistency.defer_deletes:false
2023-05-12T11:58:39.0397805Z [018] not ok 1	vinyl.select_consistency.defer_deletes:false.test_select_consistency
2023-05-12T11:58:39.0400205Z [018] #   ...tarantool/test/vinyl-luatest/select_consistency_test.lua:200: Assertion failed: 5 >= 10
2023-05-12T11:58:39.0401172Z [018] #   stack traceback:
2023-05-12T11:58:39.0402417Z [018] #   	...tarantool/test/vinyl-luatest/select_consistency_test.lua:66: in function 'run_test'
2023-05-12T11:58:39.0438606Z [018] #   	...tarantool/test/vinyl-luatest/select_consistency_test.lua:212: in function 'vinyl.select_consistency.defer_deletes:false.test_select_consistency'
2023-05-12T11:58:39.0439400Z [018] #   	...
2023-05-12T11:58:39.0440161Z [018] #   	[C]: in function 'xpcall'
2023-05-12T11:58:39.0440706Z [018] #   artifacts:
2023-05-12T11:58:39.0441957Z [018] #   	master -> /tmp/t/018_vinyl-luatest/artifacts/master-h8v2jxR4gse3
2023-05-12T11:58:39.0442635Z [018] # Starting group: vinyl.select_consistency.defer_deletes:true
2023-05-12T11:58:39.0844267Z [018] not ok 2	vinyl.select_consistency.defer_deletes:true.test_select_consistency
2023-05-12T11:58:39.1176330Z [018] #   ...tarantool/test/vinyl-luatest/select_consistency_test.lua:200: Assertion failed: 7 >= 10
2023-05-12T11:58:39.1177337Z [018] #   stack traceback:
2023-05-12T11:58:39.1178674Z [018] #   	...tarantool/test/vinyl-luatest/select_consistency_test.lua:66: in function 'run_test'
2023-05-12T11:58:39.1182316Z [018] #   	...tarantool/test/vinyl-luatest/select_consistency_test.lua:212: in function 'vinyl.select_consistency.defer_deletes:true.test_select_consistency'
2023-05-12T11:58:39.1183333Z [018] #   	...
2023-05-12T11:58:39.1184295Z [018] #   	[C]: in function 'xpcall'
2023-05-12T11:58:39.1185118Z [018] #   artifacts:
2023-05-12T11:58:39.1186294Z [018] #   	master -> /tmp/t/018_vinyl-luatest/artifacts/master-yx4w2TJGzcc0
2023-05-12T11:58:39.1187196Z [018] # Ran 2 tests in 125.260 seconds, 0 succeeded, 2 failed
2023-05-12T11:58:39.1188562Z [018] Test "vinyl-luatest/select_consistency_test.lua", conf: "None"
```

File: test/app-tap/datetime.test.lua
Comment (ligurio) [reply test/app-tap/datetime.test.lua]: Thanks!

File: src/lua/datetime.lua
Comment (ligurio) [reply src/lua/datetime.lua:1098]: Thanks!

**Review by ligurio — APPROVED** [review]
LGTM

**Comment by ligurio** [comment]
Added a label "do not merge" because PR requires rebase to master branch before merging.

**Comment by igormunkin** [comment]
Also backported to 2.11 and 2.10 in scope of daad1e7 and ca8813a respectively.

**Pull request merged**
